Simplify inactivity timeout in C#

bernardnormier opened the pull request:
This PR simplifies the inactivity timeout implementation in C#.

Once approved, I'll port it to the other languages.

bernardnormier commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,6 +1447,19 @@ internal void sendHeartbeat()
         {
             if (isActiveOrHolding())
             {
+                // We check if the connection has become inactive.
```
The check is now in sendHeartbeat, as suggested by Benoit.

bernardnormier commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,6 +1447,19 @@ internal void sendHeartbeat()
+                    _dispatchCount == 0 &&                // no pending dispatch
+                    _asyncRequests.Count == 0 &&          // no pending invocation
+                    _readHeader &&                        // we're not waiting for the remainder of an incoming message
+                    _sendStreams.Count == 0)              // no pending outgoing messages
```
We no longer check the contents of _sendStreams.

bernardnormier commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,6 +1447,19 @@ internal void sendHeartbeat()
+                    _readHeader &&                        // we're not waiting for the remainder of an incoming message
+                    _sendStreams.Count == 0)              // no pending outgoing messages
+                {
+                    scheduleInactivityTimer();
```
That's the only call to scheduleInactivityTimer.

bernardnormier commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -2024,48 +2037,22 @@ private int sendMessage(OutgoingMessage message)
-                        break; // for
-                    }
-                }
+                // The message already queued does the job of the heartbeat.
```
Interestingly, the heartbeat can still be queued (see end of method). Should we prevent this queuing too?

I think we should as it could be a "false positive": _sendStreams is not empty even though it only contains a heartbeat.

bernardnormier replied:
Actually, we should not skip a heartbeat when that's the only pending message.

bernardnormier replied:
The situation where _sendStreams contains only a heartbeat for a while is indicative of an unhealthy connection, which will be soon aborted by the idle check.

bernardnormier replied:
I am wondering if this skipping is correct:
 - we send a heartbeat to keep the connection alive
 - we find a pending message in _sendStreams so we drop it
 - the application cancels the corresponding invocation so this pending message is never sent
 - no byte gets written so the peer aborts the connection

pepone replied:
You are right if the queued invocation is about to be canceled and we don't queue the heartbeat, the peer might close the connection.

bernardnormier replied:
What should we do then?

Iterate over _sendStreams to see if there is already a non-cancellable message? The bulk of this simplification was to avoid this iteration. And checking message types would increase complexity.

cancellation messages = Request, RequestBatch?
non-cancellable message = Reply, ValidateConnection, CloseConnection?

bentoi replied:
Keep in mind that `_sendStreams.First` is the message _being_ sent not the message that will be sent next. The start of its buffer has already been written to the transport but it couldn't be fully sent and the remainder of the message will be sent by `message`.

> The situation where _sendStreams contains only a heartbeat for a while is indicative of an unhealthy connection, which will be soon aborted by the idle check.

Perhaps the transport wasn't ready to send the heartbeat (back pressure from the peer for example). It doesn't imply the connection is unhealthy. It just indicates that the sending of the heartbeat is pending. The chances of this happening with TCP are close to zero but what about other transports?

> I am wondering if this skipping is correct:
> - we send a heartbeat to keep the connection alive
> - we find a pending message in _sendStreams so we drop it
> - the application cancels the corresponding invocation so this pending message is never sent
> - no byte gets written so the peer aborts the connection

I think it's correct because the `_sendStreams.First` message is _being_ sent and its sending can't be canceled by the cancellation of the associated request.

bernardnormier replied:
> I think it's correct because the _sendStreams.First message is being sent and its sending can't be canceled by the cancellation of the associated request.

According to https://github.com/zeroc-ice/ice/blob/7c1ace30ad1cb5f39b11e9d549aaccfb3fe94e8d/csharp/src/Ice/ConnectionI.cs#L1199

_sendStreams being not empty doesn't guarantee some bytes _will_ be sent. We can be in the situation where everything has been sent and _sendStreams.First has not been removed yet.

bentoi replied:
> _sendStreams being not empty doesn't guarantee some bytes will be sent. We can be in the situation where everything has been sent and _sendStreams.First has not been removed yet.

Why does it matter here?

bentoi commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -2024,48 +2037,22 @@ private int sendMessage(OutgoingMessage message)
-            foreach (OutgoingMessage queuedMessage in _sendStreams)
+        if (_sendStreams.Count > 0)
+        {
+            if (isHeartbeat)
```
Why not just move the check to `sendHeartbeat`? Don't call `sendMessage` if `_sendStreams.Count > 0`.

bernardnormier replied:
Moved.

bernardnormier reviewed (commented):
Second attempt. Most of the logic is now in sendHeartbeat.

Note that I don't "optimize out" heartbeats. To be correct we would need to:
 - only consider messages in _sendStreams.Skip(1) since we don't know if _sendStream.First is being written or has already been written
 - check if any message in _sendStreams.Skip(1) is non-cancellable

This check seems over the top. It's just simpler to send the heartbeat no matter what.
 - optimize-out

bentoi reviewed (commented):
> Note that I don't "optimize out" heartbeats. To be correct we would need to:
> 
> - only consider messages in _sendStreams.Skip(1) since we don't know if _sendStream.First is being written or has already been written
> - check if any message in _sendStreams.Skip(1) is non-cancellable

To me, it should simply be:
```csharp
internal void sendHeartbeat()
{
    Debug.Assert(!_endpoint.datagram());

    lock (this)
    {
        if (isActiveOrHolding())
        {
            // We check if the connection has become inactive.
            if (
                _inactivityTimer is null &&           // timer not already scheduled
                _inactivityTimeout > TimeSpan.Zero && // inactivity timeout is enabled
                _state == StateActive &&              // only schedule the timer if the connection is active
                _dispatchCount == 0 &&                // no pending dispatch
                _asyncRequests.Count == 0 &&          // no pending invocation
                _readHeader &&                        // we're not waiting for the remainder of an incoming message
                _sendStreams.Count == 0               // No messages being sent
            {
                scheduleInactivityTimer();
            }

            if (_sendStreams.Count == 0)
            {
                ... send heartbeat ...
            }
            else
            {
                // No need to send a heartbeat if a message is being sent
            }
        }
        // else nothing to do
    }
}
```

In some rare cases it might not strictly be correct but it doesn't matter, we already accept a coarse grain inactivity check in favor of simpler code.

If you want to stick with more complex code, there should be a comment to explain what justifies extra checks.

pepone commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,6 +1450,39 @@ internal void sendHeartbeat()
+                    _readHeader)                          // we're not waiting for the remainder of an incoming message
+                {
+                    // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the
+                    // inactivity timer if all outgoing messages in _sendStreams are heartbeats.
```
While this might be theoretically possible it sounds like a very odd situation. Why would the peer back-pressure when there are only heartbeats?

bernardnormier replied:
The peer doesn't need a reason to back-pressure us. It can do it at any time.

bernardnormier replied:
Also, and that's very important, in this situation _sendStreams.Count will remain > 0 for a while (as long as we're back pressured). And this should not disable the inactivity timeout.

pepone commented:
Looks good, but I think we can also use the simpler version proposed by @bentoi

bernardnormier commented:
```
{
    // No need to send a heartbeat if a message is being sent
}
```

As discussed previously, this is incorrect:
 - the message "being sent" may have been sent already. We send a heartbeat to create a _new_ write. An old write is not satisfactory.
 - even when _sendStream.Count > 1, we would need to examine each message to see if it's cancellable or not. 

It's just simpler to not include this optimization.

bernardnormier commented:
> If you want to stick with more complex code, there should be a comment to explain what justifies extra checks.

There is already a comment.

bernardnormier commented:
I reworked the logic again and added comment.

The main new assumption is even if _sendStreams.First is already sent, the write occurred so recently that it's good enough to optimize-out a heartbeat.

bentoi commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,24 +1450,57 @@ internal void sendHeartbeat()
+                // a read on the peer, and resets the peer's idle check timer. When _sendStream.Count > 0, there is
+                // already an outstanding write, so we don't need to send a heartbeat. It's possible _sendStream.First
+                // was sent already but not yet removed from _sendStreams: it means the last write occurred very
+                // recently, which good enough with respect to the idle check.
```
which IS good enough

bernardnormier replied:
Fixed.

Also, I think this is good enough not only because of timing, but because of the logic in IdleTimeoutTransceiverDecorator where we schedule a new heartbeat _after_ a write completes.

When _sendStreams.First corresponds to a just-written frame, there are two possibilities:
 - the IdleTimeoutTransceiverDecorator heartbeat-rescheduling did not happen yet but is about to (=> ok to discard this heartbeat)
 - the IdleTimeoutTransceiverDecorator heartbeat-rescheduling just happened but could not cancel this heartbeat (=> ok to discard it)

bentoi commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,24 +1450,57 @@ internal void sendHeartbeat()
         {
             if (isActiveOrHolding())
```
I find it would be simpler and clearer to just remove `isActiveOrHolding` (which itself unnecessarily locks the connection). The first if block checks the state, the second could also check the state. The `idleCheck()`  implementation could do the same.

bernardnormier replied:
Ok, will remove in a follow-up PR.

bentoi reviewed (approved):
Looks good.

pepone commented on csharp/src/Ice/ConnectionI.cs:
```diff
@@ -1447,24 +1450,57 @@ internal void sendHeartbeat()
+                    // heartbeat.
+
+                    // The stream of _sendStreams.First is in _writeStream.
+                    if (_sendStreams.Count == 0 || isHeartbeat(_writeStream))
```
When does ` isHeartbeat(_writeStream)` return `True` and `_inactivityTimer` is `null`?

bernardnormier replied:
The two conditions are independent:
 - _inactivityTimer is null means the connection is active at the RPC level - for example there is an outstanding request, or an outstanding dispatch
 - isHeartbeat(_writeStream) == true means we're currently writing a heartbeat. This is an unusual situation since heartbeats are usually sent synchronously.  The peer is most likely back-pressuring us when this happens, or there is a transport issue.

pepone replied:
Say we are sending a heartbeat and `_sendStreams.Count == 0`, we schedule the inactivity timer and queue the heartbeat for sending.

If the peer is back pressuring us the heartbeat remains in `_writeStream`. 

When we are about to send the next heartbeat, the previous heartbeat remains in `_writeStream`, and we reschedule the inactivity timer.

I don't understand why the `isHeartbeat(_writeStream)` check is important here.
